rust/etw_helpers/src/sessions.rs

```diff
@@ -0,0 +1,370 @@
+use std::{borrow::Cow, mem::ManuallyDrop};
+
+use windows::{core::PCSTR, Win32::System::Diagnostics::Etw::*};
+
+#[repr(C)]
+pub struct EventTraceProperties {
+    props: EVENT_TRACE_PROPERTIES,
+    session_name: [u8; 1024],
+    file_name: [u8; 1024],
+}
+
+impl EventTraceProperties {
+    fn empty_for_query() -> EventTraceProperties {
+        unsafe {
+            let mut props: EventTraceProperties = core::mem::zeroed();
+            props.props.Wnode.BufferSize = core::mem::size_of::<Self>() as u32;
+            props.props.Wnode.Flags = WNODE_FLAG_TRACED_GUID;
+
+            props.props.LoggerNameOffset = core::mem::size_of::<EVENT_TRACE_PROPERTIES>() as u32;
+            props.props.LogFileNameOffset =
+                core::mem::size_of::<EVENT_TRACE_PROPERTIES>() as u32 + 1024;
+
+            props
+        }
+    }
+
+    pub fn set_session_name(&mut self, session_name: &str) -> &mut Self {
+        if !session_name.is_empty() {
+            unsafe {
+                let len = session_name.len();
+                if len >= 1024 {
+                    panic!()
+                }
+
+                // TODO: Can we actually use UTF-8 as a session name?
+                core::ptr::copy_nonoverlapping(
+                    session_name.as_ptr(),
+                    self.session_name.as_mut_ptr(),
+                    len,
+                );
+
+                self.session_name[len] = b'\0';
+
+                self.props.LoggerNameOffset = core::mem::size_of::<EVENT_TRACE_PROPERTIES>() as u32;
+            }
+        }
+
+        self
+    }
+
+    pub fn set_file_name(&mut self, file_name: &str) -> &mut Self {
+        if !file_name.is_empty() {
+            unsafe {
+                let len = file_name.len();
+                if len >= 1024 {
+                    panic!()
+                }
+
+                core::ptr::copy_nonoverlapping(
+                    file_name.as_ptr(),
+                    self.file_name.as_mut_ptr(),
+                    len,
+                );
+
+                self.file_name[len] = b'\0';
+
+                self.props.LogFileNameOffset =
+                    core::mem::size_of::<EVENT_TRACE_PROPERTIES>() as u32 + 1024;
+            }
+        }
+
+        self
+    }
+}
+
+pub struct ControlTraceHandle(CONTROLTRACE_HANDLE);
+
+impl Drop for ControlTraceHandle {
+    fn drop(&mut self) {
+        let mut props = EventTraceProperties::empty_for_query();
+        unsafe {
+            let ptr = &mut props.props as *mut EVENT_TRACE_PROPERTIES;
+            let _ = StopTraceA(self.0, PCSTR::null(), ptr);
+        }
+    }
+}
+
+pub struct EtwSession(core::marker::PhantomData<&'static bool>);
```

Comment:
I'm not sure I see the point of the PhantomData, or even the point of this class at all. It seems like it's just static methods that could go on a Session class.

IIRC, PhantomData is only meaningful if the class is a wrapper for data that needs some kind of lifetime tracking. Static data doesn't need lifetime tracking, so I'm confused.